src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -0,0 +1,334 @@
+
+                        isFinalBlock = EffectiveConverter.WriteCore(writer, rootValue!, Options, ref state);
+                        writer.Flush();
+
+                        bufferWriter.WriteToStream(utf8Json);
+                        bufferWriter.Clear();
+
+                        Debug.Assert(state.PendingTask == null);
+                    } while (!isFinalBlock);
+                }
+                finally
+                {
+                    if (CanUseSerializeHandler)
+                    {
+                        OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
+                    }
+                }
+            }
+        }
+
+        internal sealed override void SerializeAsObject(Utf8JsonWriter writer, object? rootValue, bool isInvokedByPolymorphicConverter = false)
+            => Serialize(writer, UnboxValue(rootValue), rootValue, isInvokedByPolymorphicConverter);
+
+        internal sealed override Task SerializeAsObjectAsync(Stream utf8Json, object? rootValue, CancellationToken cancellationToken, bool isInvokedByPolymorphicConverter = false)
+            => SerializeAsync(utf8Json, UnboxValue(rootValue), cancellationToken, rootValue, isInvokedByPolymorphicConverter);
+
+        internal sealed override void SerializeAsObject(Stream utf8Json, object? rootValue, bool isInvokedByPolymorphicConverter = false)
+            => Serialize(utf8Json, UnboxValue(rootValue), rootValue, isInvokedByPolymorphicConverter);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private T? UnboxValue(object? value)
+        {
+            if (
+#if NETCOREAPP
+                // Treated as a constant by recent versions of the JIT.
+                typeof(T).IsValueType &&
+#else
+                Type.IsValueType &&
+#endif
+                default(T) is not null && value is null)
+            {
+                // Casting null values to a non-nullable struct throws NullReferenceException, replace with JsonException
+                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type);
+            }
+
+            return (T?)value;
+        }
+
+        // Fast-path serialization in source gen has not been designed with streaming in mind.
+        // Even though it's not used in streaming by default, we can sometimes try to turn it on
+        // assuming that the current type is known to produce small enough JSON payloads.
+        // The `CanUseSerializeHandlerInStreaming` flag returns true iff:
+        //  * The type has been used in at least `MinSerializationsSampleSize` streaming serializations AND
+        //  * No serialization size exceeding JsonSerializerOptions.DefaultBufferSize / 2 has been recorded so far.
+        private bool CanUseSerializeHandlerInStreaming => _canUseSerializeHandlerInStreamingState == 1;
+        private volatile int _canUseSerializeHandlerInStreamingState; // 0: unspecified, 1: allowed, 2: forbidden
```

Comment:
Can this be simplified into a simple yes/no decision? Otherwise can we use an enum for named constants?

Reply:
not sure if compare exchange works with enums, if not you can use still use enum to assign but leave state as int (just cast it to int before assigning)